flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClientActor.java

```diff
@@ -144,11 +268,25 @@ else if (message instanceof Terminated) {
 				String msg = "Lost connection to JobManager " + jobManager.path();
 				logger.info(msg);
 				submitter.tell(decorateMessage(new Status.Failure(new Exception(msg))), getSelf());
+				resetContextAndActor();
 			} else {
 				logger.error("Received 'Terminated' for unknown actor " + target);
 			}
 		}
 
+		// ============= No messgaes received in the job manager timeout duration ========
+		else if (message instanceof ReceiveTimeout){
+			double tolerance = 0.1 * JOB_CLIENT_JOB_MANAGER_TIMEOUT.toMillis();
```

Comment:
This was to deal with the possibility that akka might enqueue a timeout message right after we get the desired message. So this checks that the timeout message is unintentional, and if we did get a message from `JobManager` just before this.